File: BlogServer/Blog.Service/Api/UserService.cs
using Blog.Model.Entity;
using Blog.Model.Param;
using Blog.Model.Rsult;
using Blog.Service.Api.Interface;
using Blog.Utils;
using Microsoft.IdentityModel.Tokens;
using Newtonsoft.Json;
using SqlSugar;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;


namespace Blog.Service.Api
{
    public class UserService : IUserService
    {
        public ISqlSugarClient Db { get; }
        public IFileService FileService { get; }

        public UserService(ISqlSugarClient sqlSugar, IFileService fileService)
        {
            Db = sqlSugar;
            FileService = fileService;
        }

        public async Task<PageResult<UserListRsult>> List(UserFindParam param)
        {
            RefAsync<int> total = 0;
            var list = await Db.Queryable<UserEnity>()
                .WhereIF(!string.IsNullOrEmpty(param.Name), it => it.Name!.Contains(param.Name!))
                .ToPageListAsync(param.Page, param.PageNum, total);
            return new PageResult<UserListRsult>
            {
                Total = total,
                List = list.Select(it => new UserListRsult(it)).ToList()
            };
        }

        public async Task<UserTokenRsult> CheckToken(string token)
        { 
            var user = await Db.Queryable<UserEnity>().Where(it => it.Token == token).FirstAsync();
            if (user == null) throw new NullReferenceException("已过期");
            user.Password = null ;
            user.Power = null ;
            return new UserTokenRsult(user);
        }

        public async Task Insert(UserInsertParam param)
        { 
            var user = await Db.Queryable<UserEnity>()
                .Where(it => it.Code == param.Code || it.Name == param.Name)
                .FirstAsync();
            if (user != null) {
                string masg = user.Code == param.Code && user.Name == param.Name ? "已存在相同的账号和名称" :
                    user.Name == param.Name ? "已存在相同的名称" : "已存在相同的账号";
                throw new InvalidOperationException(masg);
            }
            var save = new UserEnity
            {
                Name = param.Name,
                Code = param.Code,
                Status = param.Status,
                Password = GetMd5Hash.Md5HashPassword(param.Password),
                Portrait = param.Portrait != null ? JsonConvert.SerializeObject(param.Portrait) : null,
            };
            await Db.Storageable(save).ExecuteCommandAsync();
        }


        public async Task Update(UserUpdateParam param)
        {
            var userExclude = await Db.Queryable<UserEnity>()
                .Where(it => (it.Code == param.Code || it.Name == param.Name) && it.Id != param.Id)
                .FirstAsync();
            if (userExclude != null)
            {
                string masg = userExclude.Code == param.Code && userExclude.Name == param.Name ? "已存在相同的账号和名称" :
                    userExclude.Name == param.Name ? "已存在相同的名称" : "已存在相同的账号";
                throw new InvalidOperationException(masg);
            }
            var user = await Db.Queryable<UserEnity>().Where(it => it.Id == param.Id).FirstAsync();
            user.Name = param.Name;
            user.Code = param.Code;
            user.Status = param.Status;
            user.Portrait = JsonConvert.SerializeObject(param.Portrait);
            await Db.Storageable(user).ExecuteCommandAsync();
        }

        public async Task UpdatePassword(UserUpdatePasswordParam param)
        {
            var user = await Db.Queryable<UserEnity>().Where(it => it.Id == param.Id).FirstAsync();
            if (user == null) throw new NullReferenceException("找不到该用户");
            if (user.Password != GetMd5Hash.Md5HashPassword(param.Before_password)) throw new InvalidOperationException("请输入正确的密码");
            user.Password = GetMd5Hash.Md5HashPassword(param.Password);
            await Db.Storageable(user).ExecuteCommandAsync();
        }

        public async Task Delete(IDParam param)
        {
            var user = await Db.Queryable<UserEnity>().Where(it => it.Id == param.Id).FirstAsync();
            var files = JsonConvert.DeserializeObject<List<string>>(user.Portrait!)!;
            if (files.Count != 0) FileService.ImgsRemove(files);
            await Db.Deleteable<UserEnity>().In(it => it.Id,param.Id).ExecuteCommandAsync();
        }

        public async Task UpdateUserPower(UserUpdatePowerParam param)
        {
            var user = await Db.Queryable<UserEnity>().Where(it => it.Id == param.Id).FirstAsync();
            if (user == null) throw new NullReferenceException("找不到该用户");
            user.Power = JsonConvert.SerializeObject(param.Power);
            await Db.Storageable(user).ExecuteCommandAsync();
        }

        public async Task<int[]> GetUserPower(IDParam param)
        {
            var user = await Db.Queryable<UserEnity>().Where(it => it.Id == param.Id).FirstAsync();
            if (user == null) throw new NullReferenceException("找不到该用户");
            return !string.IsNullOrEmpty(user.Power) ? JsonConvert.DeserializeObject<int[]>(user.Power)! : Array.Empty<int>();
        }

        /**
         * <summary>登录</summary>
         * **/
        public async Task<SignInRsult> SignIn(SignInParam param)
        {
            UserEnity user = await Db.Queryable<UserEnity>().Where(it => it.Code == param.Code).FirstAsync();
            if (user == null) throw new NullReferenceException("查无此人");
            if (user.Password != GetMd5Hash.Md5HashPassword(param.Password)) throw new InvalidOperationException("密码错误");
            if (user.Status == 0) throw new InvalidOperationException("该账号未审核启用，请联系管理员启用");
            if (string.IsNullOrEmpty(user.Power) || user.Power.Length == 0) throw new InvalidOperationException("账号无权限，请联系管理员分配权限");
            user.Token = GenerateJwtToken(user.Code!);
            await Db.Storageable(user).ExecuteCommandAsync();
            return new SignInRsult
            {
                Token = user.Token,
                UserInfo = new UserListRsult(user)
            };

        }

        /**
         * <summary>注册token</summary>
         * **/
        public string GenerateJwtToken(string userId)
        {
            var securityKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(GlobalContext.jwtConfig!.Key));
            var credentials = new SigningCredentials(securityKey, SecurityAlgorithms.HmacSha256);
            var claims = new[]
            {
                new Claim(JwtRegisteredClaimNames.Sub, userId),
                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
            };

            var token = new JwtSecurityToken(
                issuer: GlobalContext.jwtConfig.Issuer,
                audience: GlobalContext.jwtConfig.Audience,
                claims: claims,
                expires: DateTime.Now.AddMonths(30),
                signingCredentials: credentials);

            return new JwtSecurityTokenHandler().WriteToken(token);
        }
    }

}

File: blog-server/src/module/admin/user/User.service.ts
import {ConflictException, Injectable, NotFoundException} from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm' 
import { Repository, Not, Like } from 'typeorm'
import { Md5 } from 'ts-md5/dist/md5'
import { User } from '../../../db/entities/User'
import {
    UserFindParam,
    UserInsertParam,
    UserUpdateParam,
    UserUpdatePasswordParam,
    UserUpdatePowerParam
} from "./model/User.param";
import {PageRsult} from "../../../model/BaseRsult";
import {IdParam} from "../../../model/BaseParam";
import {FileService} from "../file/file.service";
import {assignProps} from "../../../common/commonly.fun";


@Injectable()
export class UserService {
    constructor (
        @InjectRepository(User) private userRepository: Repository<User>,
        private fileService: FileService,
    ) {}
    
    // 新增用户
    async insert(body: UserInsertParam) :Promise<void>{
        const result = await this.userRepository.findOne({
            where: [{ code: body.code}, {name: body.name}]
        })
        if (result) {
            throw new ConflictException((
                result.code === body.code && result.name === body.name ? '已存在相同的账号和名称' :
                    result.name === body.name ? '已存在相同的名称' : '已存在相同的账号'
            ))
        }
        const element = new User()
        assignProps(element, body, ["password", "portrait"])
        element.password = Md5.hashStr(body.password) // 使用MD5加密
        if (body.portrait && body.portrait.length) element.portrait = JSON.stringify(body.portrait)
        await this.userRepository.save(element)
    }

    // 请求列表
    async list(body: UserFindParam): Promise<PageRsult<User>>{
        let list = await this.userRepository.find({
            where: { name: Like(`%${body.name ? body.name : ''}%`) },
            skip: body.pageNum * ( body.page - 1),
            take: body.pageNum
        })
        list = list.map(item => {
            if (item.portrait && item.portrait.length) item.portrait = JSON.parse(item.portrait)
            return item
        })
        const total = await this.userRepository.count({ where: { name: Like(`%${body.name ? body.name : ''}%`) } },)
        return {
            list,
            total
        }
    }

    // 修改用户信息
    async update(body: UserUpdateParam): Promise<void>{
        const result = await this.userRepository.findOne({ // 查询不同ID是否有相同code/name
            where: [{ code: body.code, id: Not(body.id)}, {name: body.name, id: Not(body.id)}]
        })
        if (result) {
            throw new ConflictException((
                result.code === body.code && result.name === body.name ? '已存在相同的账号和名称' :
                    result.name === body.name ? '已存在相同的名称' : '已存在相同的账号'
            ))
        }
        const user = await this.userRepository.findOne({where: { id: body.id }})

        assignProps(user, body, ["portrait"])
        if (body.portrait) user.portrait = body.portrait.length ? JSON.stringify(body.portrait) : null
        await this.userRepository.save(user)
    }

    // 修改密码
    async updatePassword(body: UserUpdatePasswordParam): Promise<void>{
        const result = await this.userRepository.findOne({
            where: { id: body.id }
        })
        if (!result) throw new NotFoundException("找不到该用户")
        if (Md5.hashStr(body.before_password) !== result.password) throw new ConflictException("请输入正确的密码")
        result.password = Md5.hashStr(body.password)
        result.token = ''
        await this.userRepository.save(result)
    }

    // 删除用户
    async delete(body: IdParam): Promise<void>{
        const result = await this.userRepository.findOne({
            where: { id: body.id }
        })
        const flies = JSON.parse(result.portrait) as Array<string>
        if (flies.length) await this.fileService.imgRemove(flies)
        await this.userRepository.delete({ id: body.id })
    }

    
    async updateUserPower(body: UserUpdatePowerParam): Promise<void>{
        const result = await this.userRepository.findOne({ 
            where: { id: body.id }
        })
        result.power = JSON.stringify(body.power)
        await this.userRepository.save(result)
    }

    async getUserPower(body:IdParam): Promise<Array<number>>{
        const result = await this.userRepository.findOne({ 
            where: { id: body.id }
        })
        return result.power && result.power.length ?  JSON.parse(result.power) as number[] : []
    }
}